### PR TITLE
server: refactor + correctness fixes for metrics

### PR DESCRIPTION
**Pull request by ngxson**
## Overview

> [!IMPORTANT]
>
> This PR corrected the way metrics are counted. The `master` branch version has some correctness issues, see below

Part of https://github.com/ggml-org/llama.cpp/issues/24866

Key changes:
- Task result reuse `stats` and `metrics` struct instead of duplicating the code
- Prompt metrics is now tied to batch --> fixed a lot of correctness issues from `master` branch
- Metrics now reflect processed and cached prompt tokens separately
- The first generated token after prompt processing isn't counted towards generation t/s. It's "free" because it comes from logits of the last prompt batch; counting it in will inflate the metrics

TODO:
- [x] identify & refactor derived metrics (for ex, t/s)
- [x] enqueue metrics updates to batch
- [x] make metrics updates tied to `llama_decode` outcome
- [x] reuse `server_slot_stats` in task result to avoid code duplication
- [x] reuse `server_metrics` in task result to avoid code duplication
- [x] refactor rendering functions

Correctness issues from `master`, fixed in this PR:
- [x] `t_prompt_processing` is N-fold counted per slot (because it's called in loop for each slot, but tokens are processed in the same batch)
- [x] First token generated is "free" but still counted towards t/s
- [x] Metrics dropped on slot fast-release paths and embedding/rerank paths
- [x] Separate prompt cached/processed metrics
- [x] Wrong stats in split prompt (multi completions from one prompt) case

## Requirements

- I have read and agree with the [contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md)
- AI usage disclosure: AI only used for validation and some mechanical/repetitive migrations

### CODE REVIEW

File: tools/server/server-context.cpp [hunk 1]
```diff
@@ -3987,6 +3884,121 @@ struct server_context_impl {
...
+        }
+
+        metrics_queue_prompt(n_prompt_tokens);
+
+        if (has_output) {
+            // sync if we have at least one output in batch
+            // so that we can calculate the timings correctly
+            llama_synchronize(ctx_tgt);
+            metrics_flush_prompt();
+        }
```
Comment (ngxson) [inline tools/server/server-context.cpp:3956]: FYI @ggerganov , we are calling `llama_synchronize` manually for batches that have at least one `output` being set

The `master` branch version accumulate the timings after sampling, which sounds better at first, but timings for prompt processing was incorrect as it ended up being counted N times more than it should be (one over-count per slot)

Comment (ngxson) [reply tools/server/server-context.cpp:3956]: One edge case: if all the decoding batch so far (async decode) are prompt, no outputs, and the slot(s) are being released (e.g. cancel request), then a sync will still be called in `callback_on_reset`

**Comment by ngxson** [comment]
I'm merging this ~tomorrow since it includes some legit fixes to the version on `master`, which currently provide wrong metrics data

CC @ServeurpersoCom if you want to double-check this

**Review by ServeurpersoCom — APPROVED** [review]
LGTM. Audited the code with a Fable loop and ran a master vs branch test on real hardware: the N-fold prompt time overcount and the free first token inflation are both confirmed on master and fixed here, the counters now match the stopwatch. Nice improvement :

```
Measurement                                    master        branch     Ground truth
predicted_per_second, n_predict=1              1000000 t/s   0.0        no decode step, rate undefined
predicted_per_second, n_predict=16             46.28 t/s     43.37 t/s  exact 16/15 ratio, free first token inflated it by 6.7%
prompt_seconds_total, 4 concurrent 4.9k req    10.3 s        5.23 s     real wall time 5.48 s, generation included
prompt throughput derived from /metrics        1904 t/s      3744 t/s   ~3600 t/s stopwatch measured
```

**Review by ggerganov — APPROVED** [review]
